File: BinarySearchStep.java
import java.io.*;
import java.util.*;

public class BinarySearchStep {
    private ArrayList<String[]> data;
    
    public int binarySearchWithSteps(String target, PrintWriter writer) {
        int left = 0;
        int right = data.size() - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            writer.println((mid+1) + ": " + data.get(mid)[0] + "/" + data.get(mid)[1]);
            
            System.out.println("Comparing target: '" + target + "' with data: '" + data.get(mid)[0] + "'");

            if (data.get(mid)[0].equals(target)) {
                writer.println("Target " + target + " found at index " + mid);
                return mid;
            } else if (data.get(mid)[0].compareTo(target) > 0) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        writer.println("-1\n");
        writer.println("Target " + target + " not found");
        return -1;
    }
    
    public void loadCSV(String filename) throws IOException {
        data = new ArrayList<>();
        BufferedReader reader = new BufferedReader(new FileReader(filename));
        String line;
        
        while ((line = reader.readLine()) != null) {
            String[] dataArray = line.split(",");
            for (int i = 0; i < dataArray.length; i++) {
                dataArray[i] = dataArray[i].trim();
            }
            data.add(dataArray);
        }
        reader.close();
    }
    
    public static void main(String[] args) throws IOException {
        if (args.length != 2) {
            System.out.println("Usage: java BinarySearchStep <dataset.csv> <target>");
            return;
        }
        
        String datasetFile = args[0];
        String target = args[1].trim(); 
        String outputFile = "binary_search_step_" + target + ".txt";
        
        BinarySearchStep searcher = new BinarySearchStep();
        searcher.loadCSV(datasetFile);
        
        if (searcher.data.isEmpty()) {
            System.out.println("Dataset is empty or invalid.");
            return;
        }
        
        System.out.println("Dataset size: " + searcher.data.size());
        System.out.println("First element: " + searcher.data.get(0)[0]);
        System.out.println("Last element: " + searcher.data.get(searcher.data.size()-1)[0]);
        System.out.println("Searching for target: " + target);
        
        PrintWriter writer = new PrintWriter(outputFile);
        int foundIndex = searcher.binarySearchWithSteps(target, writer);
        
        System.out.println("Target " + target + (foundIndex != -1 ? 
            " found at index " + foundIndex : 
            " not found") + ". Log saved to " + outputFile);
            
        writer.close();
    }
}

File: BinarySearchStep.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>

using namespace std;

int binarySearchWithSteps(const vector<vector<string>>& data, int low, int high,
                         const string& target, const string& logFile) {
    ofstream log(logFile);
    if (!log) {
        cerr << "Unable to create log file.\n";
        return -1;
    }

    while (low <= high) {
        int mid = (high + low) / 2;
        log << (mid+1) << ": " << data[mid][0] << "/" << data[mid][1] << "\n";
        
        cout << "Comparing target: '" << target << "' with data: '" << data[mid][0] << "'" << endl;

        if (data[mid][0] == target) {
            log << "Target " << target << " found at index " << mid << "\n";
            return mid;
        }

        if (data[mid][0].compare(target) > 0) {
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }
    log << "-1\n\n";
    log << "Target " << target << " not found\n";
    return -1;
}

vector<vector<string>> readDataset(const string& filename) {
    vector<vector<string>> data;
    ifstream file(filename);
    if (!file) {
        cerr << "Unable to open dataset file.\n";
        return data;
    }

    string line, word;
    while (getline(file, line)) {
        vector<string> row;
        stringstream ss(line);
        while (getline(ss, word, ',')) {
            word.erase(0, word.find_first_not_of(" \t\r\n"));
            word.erase(word.find_last_not_of(" \t\r\n") + 1);
            row.push_back(word);
        }
        data.push_back(row);
    }
    return data;
}

int main(int argc, char* argv[]) {
    if (argc != 3) {
        cerr << "Usage: " << argv[0] << " <dataset.csv> <target>\n";
        return 1;
    }

    try {
        string datasetFile = argv[1];
        string target = argv[2];
        target.erase(0, target.find_first_not_of(" \t\r\n"));
        target.erase(target.find_last_not_of(" \t\r\n") + 1);
        
        string logFile = "binary_search_step_" + target + ".txt";

        vector<vector<string>> data = readDataset(datasetFile);
        if (data.empty()) {
            cerr << "Dataset is empty or invalid.\n";
            return 1;
        }

        cout << "Dataset size: " << data.size() << endl;
        cout << "First element: " << data[0][0] << endl;
        cout << "Last element: " << data[data.size()-1][0] << endl;
        cout << "Searching for target: " << target << endl;

        int foundIndex = binarySearchWithSteps(data, 0, data.size() - 1, target, logFile);
        
        cout << "Target " << target << (foundIndex != -1 ? 
            " found at index " + to_string(foundIndex) : 
            " not found") << ". Log saved to " << logFile << "\n";
            
    } catch (const exception& e) {
        cerr << "Error: " << e.what() << "\n";
        return 1;
    }
    return 0;
}
